### ISSUE TITLE
Render Normal Map Differentiably

### ISSUE THREAD

**Issue by hzhao1997** [facebookresearch/pytorch3d#1004]
Hi, I want to optimize the mesh by calculating the L1 distance between the predicted normal map and the ground truth. So, I write a normal shader like this:
```
def NormalCalcuate(meshes, fragments):
    verts = meshes.verts_packed()  # (V, 3)
    faces = meshes.faces_packed()  # (F, 3)
    vertex_normals = meshes.verts_normals_packed()  # (V, 3)
    faces_verts = verts[faces]
    faces_normals = vertex_normals[faces]
    # pixel_coords = interpolate_face_attributes(
    #     fragments.pix_to_face, fragments.bary_coords, faces_verts
    # )
    pixel_normals = interpolate_face_attributes(
        fragments.pix_to_face, fragments.bary_coords, faces_normals #  torch.ones_like()
    )
    return pixel_normals

class NormalShader(nn.Module):
    def __init__(self, device="cuda", blend_params=None):
        super().__init__()
        self.blend_params = blend_params if blend_params is not None else BlendParams()
    def forward(self, fragments, meshes, **kwargs):
        blend_params = kwargs.get("blend_params", self.blend_params)
        # texels = meshes.sample_textures(fragments)
        normals = NormalCalcuate(meshes, fragments)
        images = softmax_rgb_blend(normals, fragments, blend_params)[:,:,:,:3]

        images = F.normalize(images, dim=3)
        return images
```
Unfortunately, it is hard for the optimized loss to decrease. Is there any wrong with my shader? Could you give me some advice? Thanks a lot : )

**Comment by timlod** [comment]
From my experience, shader output is not differentiable with respect to the mesh, ie. RGB loss based on `SoftPhongShader` output.
I'd assume the same holds for a normal map. You can differentiate wrt. the normal map itself to match a target, but you can't change the mesh from the resulting loss.
https://github.com/facebookresearch/pytorch3d/issues/839 suggests this as well (although `SilhouetteShader` seems to work).
I'm not yet sure if this is a bug - should it be possible to deform a mesh based on all rendering output, or not?

**Comment by gkioxari** [comment]
@timlod The idea of differentiable rendering is that gradients propagate back from the rendered RGB image to the 3D shape.  So yes, you can deform a mesh based on losses on the output RGB images. Please check out our tutorials https://github.com/facebookresearch/pytorch3d/blob/main/docs/tutorials/fit_textured_mesh.ipynb

@hzhao1997 Your question is more of a modeling question. If you believe gradients are not propagated you can check the `.grad` field of your variables. You can move this into the GH Discussions if you wish others to help you out, but this is not a bug with the library.

**Comment by timlod** [comment]
@gkioxari Sorry, I forgot to update here. Indeed, it works (I was able to get there after revisiting my problem from scratch and a lot of tweaking).
I was thinking there may be something wrong because of many other issues that could not get any/good results after rendering (see the one I mentioned and e.g. #774, #359). I think it comes down to the problem being quite fickle, and the setup usually being underconstrained - gradients get noisy and optimization of the mesh seems impossible.

I imagine that OP similarly may have issues due to not constraining the problem enough. Or possibly, that they imagine a loss against a texture normal map, which may not work because pytorch3d normals are interpolated face normals (perhaps it could still work with a very high-res mesh). I have implemented normal mapping in my application and will post some examples at #642.
